lib/auth/machineid/machineidv1/spiffe_federation_syncer.go

```diff
@@ -0,0 +1,631 @@
+	defer func() {
+		wg.Wait()
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Set up state management that will hold a list of all active trust domain
+	// syncers.
+	states := map[string]trustDomainSyncState{}
+	mu := &sync.Mutex{}
+	startSyncingFederation := func(trustDomain string) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		// Don't start if we're already syncing this trust domain.
+		_, ok := states[trustDomain]
+		if ok {
+			return
+		}
+
+		eventsCh := make(chan types.Event)
+		states[trustDomain] = trustDomainSyncState{
+			eventsCh: eventsCh,
+		}
+
+		wg.Add(1)
+		go func() {
+			defer func() {
+				mu.Lock()
+				delete(states, trustDomain)
+				mu.Unlock()
+				wg.Done()
+			}()
+			s.syncTrustDomainLoop(ctx, trustDomain, eventsCh)
+		}()
+	}
+
+	// Establish our watcher, we'll use this to react instantly to changes to SPIFFEFederations.
+	w, err := s.cfg.EventsWatcher.NewWatcher(ctx, types.Watch{
+		Kinds: []types.WatchKind{{
+			Kind: types.KindSPIFFEFederation,
+		}},
+	})
+	if err != nil {
+		return trace.Wrap(err, "failed to create watcher")
+	}
+	defer func() {
+		err := w.Close()
+		if err != nil {
+			s.cfg.Logger.ErrorContext(
+				ctx, "Failed to close watcher", "error", err,
+			)
+		}
+	}()
+
+	// Wait for initial "Init" event to indicate we're now receiving events.
+	select {
+	case <-w.Done():
+		if err := w.Error(); err != nil {
+			return trace.Wrap(err, "watcher failed")
+		}
+		return trace.BadParameter("watcher closed unexpectedly")
+	case evt := <-w.Events():
+		if evt.Type == types.OpInit {
+			break
+		}
+		return trace.BadParameter("expected init event, got %v", evt.Type)
+	case <-ctx.Done():
+		return nil
+	}
+
+	// Fetch an initial list of all federations and start syncers for them.
+	trustDomains, err := listAllTrustDomains(ctx, s.cfg.Store)
+	if err != nil {
+		return trace.Wrap(err, "initially listing trust domains")
+	}
+	for _, td := range trustDomains {
+		startSyncingFederation(td.GetMetadata().GetName())
+	}
+
+	// Now we can start reacting to events, we'll want to start/stop syncers as
+	// needed. We'll want to start a syncer for any new trust domain, and
+	// propagate events to existing syncers.
+	for {
+		select {
+		case evt := <-w.Events():
+			s.cfg.Logger.DebugContext(
+				ctx,
+				"Received event from SPIFFEFederation watcher",
+				"evt_type", evt.Type,
+			)
+			if !(evt.Type == types.OpPut || evt.Type == types.OpDelete) {
+				continue
+			}
+
+			mu.Lock()
+			existingState, ok := states[evt.Resource.GetName()]
+			mu.Unlock()
+
+			// If it already exists, we can just pass the event along. If
+			// there's already a sync queued due to an event, we don't need to
+			// queue another since it fetches the latest resource anyway.
+			if ok {
+				select {
+				case existingState.eventsCh <- evt:
+				default:
+					s.cfg.Logger.DebugContext(
+						ctx,
+						"Sync already queued for trust domain, ignoring event",
+					)
+				}
+				continue
+			}
```

Comment:
I've patched this by handling the deletion within the watcher rather than within the loop. If an OpDelete comes through followed shortly put a OpPut, the original syncer will stop and a new syncer will replace it.